Changes to the MD tutorial for the 2020 course

AfroditiMariaZaki opened the pull request:
Updates on MD simulation:
-  Done with the GMX version2020 
-   mdp files updated
-   forcefield changed to AMBER99SB-ILDN
-   water model changed to TIP3P
-   submission scripts added for ARCUS

Added 5 jupyter notebooks that contain the different sections of the tutorial

AfroditiMariaZaki committed:
Changes to the MD tutorial for the 2020 course

IAlibay commented on tutorials/MD/archive-2019/CONTRIBUTORS.md:
```diff
@@ -0,0 +1,9 @@
+# List of contributors to the Oxford Computational Biochemitry Molecular Dynamics course
```
I think can we can ahead and just delete all the archive-2019 stuff. It's already in the 2019 release, so there's a git memory of it.

philbiggin replied:
Yes - please delete.

AfroditiMariaZaki replied:
Deleted.

IAlibay commented:
@AfroditiMariaZaki thanks for creating the PR.

Just as a quick first comment, unless @philbiggin has any objections, can you just go ahead and delete all the archive-2019 stuff? It's already in the `v2019` release, so there's no advantage to keeping it around in `master`. If anything it'll make review easier to do.

IAlibay commented on tutorials/MD/data/md.mdp:
```diff
@@ -1,314 +1,45 @@
+continuation            = no        ; Not continuing from previous simulation 
+constraint_algorithm    = lincs     ; holonomic constraints 
+constraints             = h-bonds   ; bonds involving H are constrained
+lincs_iter              = 1         ; accuracy of LINCS
```
I believe the modern gmx way to handle these variables is to use a `-` instead of a `_`, it doesn't have any impact on 2020, but it could do on future versions. It might be safer to change these here where possible to avoid having to think about it in future releases?

I've added comment suggested changes so as to speed up the process.

AfroditiMariaZaki replied:
_ changed to - in .mdp files

IAlibay commented on tutorials/MD/data/md.mdp:
```diff
@@ -1,314 +1,45 @@
+compressed-x-grps       = System    ; save the whole system
+; Bond parameters
+continuation            = no        ; Not continuing from previous simulation 
+constraint_algorithm    = lincs     ; holonomic constraints 
```
```suggestion
constraint-algorithm    = lincs     ; holonomic constraints 
```

IAlibay commented on tutorials/MD/data/md.mdp:
```diff
@@ -1,314 +1,45 @@
+continuation            = no        ; Not continuing from previous simulation 
+constraint_algorithm    = lincs     ; holonomic constraints 
+constraints             = h-bonds   ; bonds involving H are constrained
+lincs_iter              = 1         ; accuracy of LINCS
```
```suggestion
lincs-iter              = 1         ; accuracy of LINCS
```

IAlibay commented on tutorials/MD/data/md.mdp:
```diff
@@ -1,314 +1,45 @@
+constraint_algorithm    = lincs     ; holonomic constraints 
+constraints             = h-bonds   ; bonds involving H are constrained
+lincs_iter              = 1         ; accuracy of LINCS
+lincs_order             = 4         ; also related to accuracy
```
```suggestion
lincs-order             = 4         ; also related to accuracy
```

IAlibay commented on tutorials/MD/data/md.mdp:
```diff
@@ -1,314 +1,45 @@
+rvdw                    = 1.0       ; short-range van der Waals cutoff (in nm)
+; Electrostatics
+coulombtype             = PME       ; Particle Mesh Ewald for long-range electrostatics
+pme_order               = 4         ; cubic interpolation
```
```suggestion
pme-order               = 4         ; cubic interpolation
```

AfroditiMariaZaki replied:
Fixed

IAlibay commented on tutorials/MD/data/md.mdp:
```diff
@@ -1,314 +1,45 @@
+lincs_order             = 4         ; also related to accuracy
+; Neighborsearching
+cutoff-scheme           = Verlet    ; Buffered neighbor searching
+ns_type                 = grid      ; search neighboring grid cells
```
```suggestion
ns-type                 = grid      ; search neighboring grid cells
```

AfroditiMariaZaki replied:
Fixed

IAlibay commented on tutorials/MD/data/md.mdp:
```diff
@@ -1,314 +1,45 @@
+; Pressure coupling is on
+pcoupl                  = Parrinello-Rahman     ; Pressure coupling on in NPT
+pcoupltype              = isotropic             ; uniform scaling of box vectors
+tau_p                   = 1.0                   ; time constant, in ps
```
```suggestion
tau-p                   = 1.0                   ; time constant, in ps
```

AfroditiMariaZaki replied:
Fixed

IAlibay commented on tutorials/MD/data/md.mdp:
```diff
@@ -1,314 +1,45 @@
+pcoupl                  = Parrinello-Rahman     ; Pressure coupling on in NPT
+pcoupltype              = isotropic             ; uniform scaling of box vectors
+tau_p                   = 1.0                   ; time constant, in ps
+ref_p                   = 1.0                   ; reference pressure, in bar
```
```suggestion
ref-p                   = 1.0                   ; reference pressure, in bar
```

AfroditiMariaZaki replied:
Fixed

IAlibay commented on tutorials/MD/data/md.mdp:
```diff
@@ -1,314 +1,45 @@
+; Dispersion correction
+DispCorr                = EnerPres   ; account for cut-off vdW scheme
+; Velocity generation
+gen_vel                 = yes        ; Velocity generation is on
```
```suggestion
gen-vel                 = yes        ; Velocity generation is on
```

AfroditiMariaZaki replied:
Fixed

IAlibay commented on tutorials/MD/data/md.mdp:
```diff
@@ -1,314 +1,45 @@
+tcoupl                  = v-rescale             ; modified Berendsen thermostat
+tc-grps                 = System                ; one coupling group
+tau_t                   = 0.5                   ; time constant, in ps
+ref_t                   = 300                   ; reference temperature, one for each group, in K
```
```suggestion
ref-t                   = 300                   ; reference temperature, one for each group, in K
```

AfroditiMariaZaki replied:
Fixed

IAlibay commented on tutorials/MD/data/md.mdp:
```diff
@@ -1,314 +1,45 @@
+; Temperature coupling is on
+tcoupl                  = v-rescale             ; modified Berendsen thermostat
+tc-grps                 = System                ; one coupling group
+tau_t                   = 0.5                   ; time constant, in ps
```
```suggestion
tau-t                   = 0.5                   ; time constant, in ps
```

IAlibay commented on tutorials/MD/data/submit_em.sh:
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
```
Let's see what ARC returns with, we might have to call the GPU partition here.

AfroditiMariaZaki replied:
Yes, submission scripts updated

IAlibay commented on tutorials/MD/prerun/run/em.log:
```diff
@@ -1,22 +1,22 @@
-Log file opened on Sat Dec 14 15:07:00 2019
-Host: diablo  pid: 17682  rank ID: 0  number of ranks:  1
-                      :-) GROMACS - gmx mdrun, 2018.4 (-:
+                      :-) GROMACS - gmx mdrun, 2020.1 (-:
```
Is there a point in storing the log files?

It's a very large file for github to track, maybe if we just want to make it available for those that can't use arcus we can zip the non-system (i.e. not a trajectory/structure/etc... file)?

IAlibay replied:
same thing for the trr file too (unless we plan on using it?)

AfroditiMariaZaki replied:
.log and .trr files deleted

IAlibay commented on tutorials/MD/prerun/setup/em.mdp:
```diff
@@ -0,0 +1,14 @@
+; Parameters describing how to find the neighbors of each atom and how to calculate the interactions
+nstlist         = 1         ; Frequency to update the neighbor list and long range forces
+cutoff-scheme   = Verlet    ; Buffered neighbor searching
+ns_type         = grid      ; Method to determine neighbor list (simple, grid)
```
```suggestion
ns-type         = grid      ; Method to determine neighbor list (simple, grid)
```

IAlibay commented on tutorials/MD/prerun/setup/setup/em.mdp:
```diff
@@ -0,0 +1,14 @@
+; Parameters describing how to find the neighbors of each atom and how to calculate the interactions
+nstlist         = 1         ; Frequency to update the neighbor list and long range forces
+cutoff-scheme   = Verlet    ; Buffered neighbor searching
+ns_type         = grid      ; Method to determine neighbor list (simple, grid)
```
```suggestion
ns-type         = grid      ; Method to determine neighbor list (simple, grid)
```

setup in setup? Is this intended?

IAlibay reviewed (changes requested):
Just some initial first pass comments, I'll have a deeper look at the notebooks tomorrow. Also my bad on the .npz file.

I've added the file type to the gitignore, so it should be safe to remove now.

IAlibay commented on tutorials/MD/01_Introduction.ipynb:
```diff
@@ -0,0 +1,64 @@
+    "\n",
+    "It is divided into five  parts, i.e, five jupyter notebooks. \n",
+    "\n",
+    "* **00_**\n",
```
Let's call this one "setup"

AfroditiMariaZaki replied:
Done

IAlibay commented on tutorials/MD/01_Introduction.ipynb:
```diff
@@ -0,0 +1,64 @@
+    "\n",
+    "* **00_**\n",
+    "\n",
+    "Hopefully, you have already gone through this notebook, as it essential for everything that follows. You should have already installed all the software packages that you need for the tutorial and are familiar with how to use the RCUS cluster and how to transfer your files to your laptop. If you haven't already read it, please go back and make sure everything is setup!\n",
```
RCUS -> ARCUS

AfroditiMariaZaki replied:
Corrected

IAlibay replied:
Still showing as RCUS in this draft PR?

AfroditiMariaZaki replied:
Updated again, should be ok now

IAlibay commented on tutorials/MD/01_Introduction.ipynb:
```diff
@@ -0,0 +1,64 @@
+{
```
Rather than a notebook, would this be better handled as just a README.md file?

It feels like this information should come before the "setup" instructions.

IAlibay replied:
Ignore this if you haven't done so yet, let's go full notebook (it's simpler)

IAlibay commented on tutorials/MD/01_Introduction.ipynb:
```diff
@@ -0,0 +1,64 @@
+    "\n",
+    "* **04_Trajectory_Analysis** \n",
+    "\n",
+    "The final notebook will hopefully show you some types of information that we can gain from a MD trajectory, show you how to use MDAnalysis and Pyplot to obtain and plot the data and will encourage you to think about how to interpret the data that you have obtained. \n",
```
Can the references to the software be links? That way folks could click on them and see where they can get more information if they want to.

AfroditiMariaZaki replied:
Done

IAlibay commented on tutorials/MD/02_Protein_Visualisation.ipynb:
```diff
@@ -0,0 +1,55 @@
+{
```
So given what we do with VMD in this notebook and 04, it could be possible to ditch it completely, however I realise that there might be some interest in getting folks using VMD too.

My suggestion here is to use nglview as the primary visualisation method and then have VMD as an optional backup. If they are interested they can do both, but otherwise if they happen to have issues setting things up, they can just boot up binder or something and just use nglview.

For nglview all we would need to do is:

```python
import nglview as nv
pdbview = nv.show_pdbid('1HSG')
pdbview.display(gui=True)
```

This will generate an NGLView output.

We then say "nglview is great for looking at things directly within a jupyter notebook, but VMD can be a more power/intuitive tool, we optionally recommend that you look at this protein with VMD etc..."

hainm replied:
@IAlibay  Sorry to jump into this PR (from my search for nglview):

> pdbview.display(gui=True)

can be replaced by

> pdbview.gui_style = 'ngl'
https://github.com/nglviewer/nglview/pull/830

IAlibay replied:
Thanks @hainm, that's really useful to know :)

P.S. 

> but VMD can be a more power/intuitive tool

I only mean this in the strictest "single non-pythonic tool", nglview+MDAnalysis is still my favourite ;)

AfroditiMariaZaki replied:
Changed to include both a VMD and an NGLView section

IAlibay commented on tutorials/MD/03_MD_Setup_Production_Run.ipynb:
```diff
@@ -0,0 +1,274 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Connect to the Advanced Research Computing (ARC)\n",
```
We should have consistency with major titles like "03: Running an MD simulation"?

AfroditiMariaZaki replied:
Fixed the titles and subtitles.

IAlibay commented on tutorials/MD/03_MD_Setup_Production_Run.ipynb:
```diff
@@ -0,0 +1,274 @@
+    "\n",
+    "Open a terminal and remotely connect to ARCUS using ssh (secure shell), using your username:\n",
+    "\n",
+    "`% ssh -X username@oscgate.arc.ox.ac.uk`\n",
```
You're not using X-forwarding in the tutorial, can we drop the `-X` flags?

AfroditiMariaZaki replied:
Done

IAlibay commented on tutorials/MD/03_MD_Setup_Production_Run.ipynb:
```diff
@@ -0,0 +1,274 @@
+    "\n",
+    "`% ssh -X arcus-htc`\n",
+    "\n",
+    "If you get a message like this one:\n",
```
Let's remove these details from here and move them to 00_setup. This won't happen if they log in again and it may just lead to confusion.

Edit; here I mean specifically the RSA key bit.

AfroditiMariaZaki replied:
RSA bit removed. I kept the login instructions.

IAlibay commented on tutorials/MD/03_MD_Setup_Production_Run.ipynb:
```diff
@@ -0,0 +1,274 @@
+    "`% mkdir setup`\\\n",
+    "`% cd setup`\n",
+    "\n",
+    "Now, you need to load GROMACS, which is the simulation software package that you will utilise:\n",
```
Can we move this under the next section, renaming the next section as "The GROMACS Molecular Dynamics Engine"?

That way you have this nice introduction that says "this is what you'll be using", and then you can go "in ARCUS to use gromacs you should do the following".

AfroditiMariaZaki replied:
Done

IAlibay commented on tutorials/MD/03_MD_Setup_Production_Run.ipynb:
```diff
@@ -0,0 +1,274 @@
+   "source": [
+    "## Specifics to GROMACS\n",
+    "\n",
+    "GROMACS is a versatile package to perform molecular dynamics, i.e. simulate the Newtonian equations of motion for systems with hundreds to millions of particles. (http://www.gromacs.org/)\n",
```
Maybe worth an extra sentence here to say that a) it is free, b) it is what we will be using here, c) there's a lot of options/functionality that we will not be covering here. "This tutorial will only scratch the surface of all the powerful things GROMACS can do" or something like this.

AfroditiMariaZaki replied:
Done

IAlibay commented on tutorials/MD/03_MD_Setup_Production_Run.ipynb:
```diff
@@ -0,0 +1,274 @@
+    "\n",
+    "GROMACS is a versatile package to perform molecular dynamics, i.e. simulate the Newtonian equations of motion for systems with hundreds to millions of particles. (http://www.gromacs.org/)\n",
+    "\n",
+    "- Coordinates files have the extension .gro and the default name is conf.gro.\n",
```
This should be prefaced by a header maybe? Something like "A brief overview of GROMACS file types".

With maybe a follow on line going "Gromacs supports many different file types, here we give a brief overview of the different file types we will encounter during this tutorial".

AfroditiMariaZaki replied:
Done

IAlibay commented on tutorials/MD/03_MD_Setup_Production_Run.ipynb:
```diff
@@ -0,0 +1,274 @@
+    "\n",
+    "- Coordinates files have the extension .gro and the default name is conf.gro.\n",
+    "- The topology file (default name topol.top) contains all the information about which atoms are bonded to which and what force-field parameters are applied etc.\n",
+    "- The trajectory files have the extension .xtc and .trr, the former does not contain velocity information and coordinates are held at a reduced precision, and so occupies less disk space. However you will need velocities if you want to continue a simulation.\n",
```
Should we also through in .tng here? Technically it is the format we should be using, but adoption has been very slow. That being said one of our GSOC students has gone a long way into making it more accessible.

AfroditiMariaZaki replied:
Haven't used it throughout the tutorial and I only discussed files that are used. If you think it is necessary, I can add a sentence.

IAlibay commented on tutorials/MD/03_MD_Setup_Production_Run.ipynb:
```diff
@@ -0,0 +1,274 @@
+    "- The trajectory files have the extension .xtc and .trr, the former does not contain velocity information and coordinates are held at a reduced precision, and so occupies less disk space. However you will need velocities if you want to continue a simulation.\n",
+    "- The .edr file contains the energy information from the trajectory.\n",
+    "- The .mdp file contains the information that was used to setup the actual simulation. Things related to temperature, pressure, how the electrostatics is calculated etc. Although we will not use this today, the file is provided so you can see how this trajectory was made.\n",
+    "- The .ndx file allows you to specify atoms or groups of atoms for use in analysis, restraints, etc and is optional.\n",
```
Do we generate ndx files anywhere in this tutorial?

AfroditiMariaZaki replied:
No, you are right. We did in the previous version for HB analysis. I deleted it.

IAlibay commented on tutorials/MD/03_MD_Setup_Production_Run.ipynb:
```diff
@@ -0,0 +1,274 @@
+    "- The topology file (default name topol.top) contains all the information about which atoms are bonded to which and what force-field parameters are applied etc.\n",
+    "- The trajectory files have the extension .xtc and .trr, the former does not contain velocity information and coordinates are held at a reduced precision, and so occupies less disk space. However you will need velocities if you want to continue a simulation.\n",
+    "- The .edr file contains the energy information from the trajectory.\n",
+    "- The .mdp file contains the information that was used to setup the actual simulation. Things related to temperature, pressure, how the electrostatics is calculated etc. Although we will not use this today, the file is provided so you can see how this trajectory was made.\n",
```
The last part of this sentence is confusing, technically we do use the mdp right? Maybe it should be "although we will not be exploring it in details today, example mdp files are included". We should probably put a disclaimer that the settings we use here are not optimal and show only be used for illustration purposes.

AfroditiMariaZaki replied:
Yes, we did use the .mdp, I removed that sentence.

IAlibay commented on tutorials/MD/03_MD_Setup_Production_Run.ipynb:
```diff
@@ -0,0 +1,274 @@
+    "\n",
+    "In this section, we will obtain our protein coordinates and perform some routine Molecular Dynamics calculations on them.\n",
+    "\n",
+    "For this tutorial, we will use the HIV-1 protease structure (1HSG). It is a homodimer with two chains of 99 residues each. The .pdb file of the protein can be obtained from the protein data bank (https://www.rcsb.org/), but it is probably easier for now to just copy it from the shared directory to your working directory. In fact, most of the files you need can be found in this directory so i ...
```
It may be best here to link back to the previous notebook, say something like "as previously described in notebook 02, we will be simulating the structure of HV-1 protease...". And then mention how they could have uploaded the PDB files through sftp, but we did this for them already.

AfroditiMariaZaki replied:
Done

IAlibay commented on tutorials/MD/03_MD_Setup_Production_Run.ipynb:
```diff
@@ -0,0 +1,274 @@
+    "\n",
+    "Use your preferred text editor to open the protein.pdb file and see how it differs from the 1hsg.pdb file.\n",
+    "\n",
+    "We now need to make sure that all the hydrogens are added to our protein. This process will also generate the\n",
```
It might be more easy to understand if this sentence is reversed, i.e. generate the parameters, and will also add hydrogens where necessary (should we also detail what -ignh does)?

AfroditiMariaZaki replied:
Done

IAlibay commented on tutorials/MD/03_MD_Setup_Production_Run.ipynb:
```diff
@@ -0,0 +1,274 @@
+    "\n",
+    "`% gmx editconf -f protein.gro -box 7 7 7 -c -o boxed.gro`\n",
+    "\n",
+    "This puts the protein in the centre of the box that is 7 nm x 7 nm x 7 nm and creates the resulting file boxed.pdb .\\\n",
```
Do we need a quick one liner about periodic boundaries here?

AfroditiMariaZaki replied:
Done, even though this will also be discussed during the lecture.

IAlibay commented on tutorials/MD/03_MD_Setup_Production_Run.ipynb:
```diff
@@ -0,0 +1,274 @@
+    "\n",
+    "`% gmx editconf -f protein.gro -box 7 7 7 -c -o boxed.gro`\n",
+    "\n",
+    "This puts the protein in the centre of the box that is 7 nm x 7 nm x 7 nm and creates the resulting file boxed.pdb .\\\n",
```
Extra `\` at the end of the line with "boxed.pdb"., Also this is a gro file?

AfroditiMariaZaki replied:
\ removed, corrected to read boxed.gro

IAlibay commented on tutorials/MD/03_MD_Setup_Production_Run.ipynb:
```diff
@@ -0,0 +1,274 @@
+    "`% gmx editconf -f protein.gro -box 7 7 7 -c -o boxed.gro`\n",
+    "\n",
+    "This puts the protein in the centre of the box that is 7 nm x 7 nm x 7 nm and creates the resulting file boxed.pdb .\\\n",
+    "Next we need to add water to the system. We will also add ions -enough to neutralize the system and to a\n",
```
The adding ions here is a bit messy because we do the solvation and then we then again have a blurb about adding ions. Would it make sense to merge both and only mention it after the solvate call?

AfroditiMariaZaki replied:
Rephrased it.

IAlibay commented on tutorials/MD/03_MD_Setup_Production_Run.ipynb:
```diff
@@ -0,0 +1,274 @@
+    "\n",
+    "### Energy Minimization\n",
+    "\n",
+    "Before we can run the actual dynamics, we need to first minimize the energy of the system. Ideally you would minimize down until the forces were below a certain level (tolerance), but we will just give a quick burst of 200 steps here. Since we have finished setting up the system, we will now move to a new directory to perform our simulation from:\n",
```
It might be good to add a line here to mention why we do this (i.e. maybe mention that water & hydrogen addition is never ideal, neither are input crystal structures).

AfroditiMariaZaki replied:
An extra sentence explaining the need for EM added.

IAlibay commented on tutorials/MD/03_MD_Setup_Production_Run.ipynb:
```diff
@@ -0,0 +1,274 @@
+    "\n",
+    "`% cd run`\n",
+    "\n",
+    "The `grompp` command will read the information of the system that we will provide (coordinates, topologies and simulation parameters) and will generate a run input file:\n",
```
`.tpr run input file`

AfroditiMariaZaki replied:
Changed.

IAlibay commented on tutorials/MD/03_MD_Setup_Production_Run.ipynb:
```diff
@@ -0,0 +1,274 @@
+    "\n",
+    "`% cp /home/shared/data/submit_em.sh .`\n",
+    "\n",
+    "You can explore its contents and see that the last line in the file contains the `mdrun` command by typing: \n",
```
Maybe worth adding an extra line about what `mdrun` does, i.e. it invokes the main MD engine bits of gromacs?

AfroditiMariaZaki replied:
Done

IAlibay commented on tutorials/MD/03_MD_Setup_Production_Run.ipynb:
```diff
@@ -0,0 +1,274 @@
+    "\n",
+    "It will take a few minutes to run, depending on the waiting times of the queue. Check on the status of the run by typing:\n",
+    "\n",
+    "`% squeue -u username`\n",
```
can we just use `$USER` here to avoid them asking what their username is?

AfroditiMariaZaki replied:
Done, even thought the username will be needed later for the scp command as well? and it appears at the beginning of their command line I think.

IAlibay commented on tutorials/MD/03_MD_Setup_Production_Run.ipynb:
```diff
@@ -0,0 +1,274 @@
+    "\n",
+    "Remember to replace `username` with your own username! It should print something like this:\n",
+    "\n",
+    "`             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON) \n",
```
The alignment here isn't great on my end, are you getting the correct alignment?

AfroditiMariaZaki replied:
Tried to fix that, I think it is better now.

IAlibay commented on tutorials/MD/03_MD_Setup_Production_Run.ipynb:
```diff
@@ -0,0 +1,274 @@
+    "\n",
+    "`% tail -n 12 em.log`\n",
+    "\n",
+    "This will show you what step the simulation is at or if it has finished.\\\n",
```
Can we make them look for the gromacs quote as an indicator that the job has finished properly?

AfroditiMariaZaki replied:
It is not printed at the end of the .log file

IAlibay commented on tutorials/MD/03_MD_Setup_Production_Run.ipynb:
```diff
@@ -0,0 +1,274 @@
+    "type in 10 when prompted (should correspond to potential energy from the list of options presented), then a zero to finish.\\\n",
+    "The plot should look like this one:\n",
+    "\n",
+    "<img src=\"em_potential_energy.png\" width=500 height=400/>\n",
```
This isn't present (also we aren't plotting here?) Maybe we can just say that we will be doing the plotting in section 04 but that usually you would do this before moving on to an MD simulation?

IAlibay replied:
We could get them to vi/nano/etc... the xvg file and look at the decrease in energy maybe?

AfroditiMariaZaki replied:
Image removed. Instead, they are asked to open the file with the text editor to observe the decrease in energy.

IAlibay commented on tutorials/MD/03_MD_Setup_Production_Run.ipynb:
```diff
@@ -0,0 +1,274 @@
+    "\n",
+    "### Production Run\n",
+    "\n",
+    "At this stage, we would normally run a short simulation where the protein atoms are restrained while the water molecules\n",
```
Can we point them to the md tutorials website for a more detailed step by step process where they do this?

Also do we need to explain why we would do this normally?

AfroditiMariaZaki replied:
Briefly explained. Will be discussed during lecture too

IAlibay commented on tutorials/MD/03_MD_Setup_Production_Run.ipynb:
```diff
@@ -0,0 +1,274 @@
+    "At the moment it is set up to run for 1000 ps. This will take several minutes to complete depending on the waiting time in the queue - time for lunch! You don't have to wait for it to finish completely though, although now might be a good time for a break to allow at least some data to appear. The analysis can be done on the output files that will be generated or you can always use the \"one ...
+    "\n",
+    "After the end of the production run, it would be useful to obtain some properties that will give us an insight into our protein system.\\\n",
+    "There are various so-called ensembles that are used for protein simulations - probably the most common is a system where the number of particles, the pressure and the temperature are held constant (NPT). This is usually achieved by means of a heat-bath. Nevertheless, it is usually a good idea to check these as a function of time through the trajectory just to make sure nothing unexpected hap ...
```
The "This is usually achieved by means of a heat-bath. " just seems a little bit out of place. Can we just say something like "as detailed in the .mdp file, here we use the V-rescale thermostat (link to paper), which controls the simulation temperature through the use of a heat bath coupled with a stochastic rescaling algorithm. Whilst it is quite a robust algorithm, it is usually a good..."

AfroditiMariaZaki replied:
More info on v-rescale added.

IAlibay commented on tutorials/MD/03_MD_Setup_Production_Run.ipynb:
```diff
@@ -0,0 +1,274 @@
+    "\n",
+    "We shall select short-range lennard-jones (7), short range coulombic (9) and the potential energy (11). End your selection with a zero.\n",
+    "\n",
+    "Finally, we need to renumber the em.gro file so that the residues of the two chains will not have the same residue number. This will be necessary for when we calculate the RSF values.\n",
```
RSF -> RMSF?

Also can we not just do this directly within MDAnalysis by using the residue index?

AfroditiMariaZaki replied:
Yes, removed

IAlibay commented on tutorials/MD/03_MD_Setup_Production_Run.ipynb:
```diff
@@ -0,0 +1,274 @@
+    "\n",
+    "The use `scp` to transfer the remote run subdirectory to your local directory:\n",
+    "\n",
+    "`% scp username@oscgate.arc.ox.ac.uk:/home/username/run/ .`\n",
```
We probably need to be clearer as to where they are copying to (i.e. the place where the notebooks exist)?

AfroditiMariaZaki replied:
Clarified it.

IAlibay commented on tutorials/MD/04_Trajectory_Analysis.ipynb:
```diff
@@ -0,0 +1,265 @@
+   "source": [
+    "### Visualise the simulation\n",
+    "\n",
+    "The simplest and easiest type of analysis you should always do is to look at it with your eyes! Your eyes will tell if you something strange is happening immediately. A numerical analysis may not. Let us look at the simulations on VMD.\n",
```
Again here I propose we make VMD optional and have nglview as the first option.

here it should just be a case of calling MDAnalysis, creating a universe, getting the protein atomgroup out and then passing that to nglview , as detailed here: http://nglviewer.org/nglview/latest/api.html#nglview.show_mdanalysis

AfroditiMariaZaki replied:
NGLView instructions included. VMD reminder that it is optional too.

IAlibay commented on tutorials/MD/03_MD_Setup_Production_Run.ipynb:
```diff
@@ -0,0 +1,274 @@
+    "\n",
+    "`% gmx editconf -f em.gro -o em_renumbered.gro -resnr 1`\n",
+    "\n",
+    "We will plot and explore the temperature and the energetic components that we obtained in the next section of the tutorial.\n",
```
Might have missed it somewhere, but we should get them to trjconv their system to make sure it stayed in the box? It's unlikely to have moved that much in 1 ns, but still possible. If we don't do this here, we'll have to do it later using MDAnalysis, which is also ok to do, but a little bit more complex in terms of number of python lines.

AfroditiMariaZaki replied:
Updated. They are asked to modify the trajectory to place the protein at the center and remove rotational and translational motion.

IAlibay commented on tutorials/MD/04_Trajectory_Analysis.ipynb:
```diff
@@ -0,0 +1,265 @@
+    "rmsf_calc = rmsf(calphas, verbose=True).run()\n",
+    "\n",
+    "#Plot the RMSF\n",
+    "pyplot.plot(calphas.resids, rmsf_calc.rmsf, color='darkorange' )\n",
```
rather than residues, you should be able to use resindices+1 or maybe resindex, to get a "continuous residue number starting from 1"?

AfroditiMariaZaki replied:
Fixed

IAlibay reviewed (changes requested):
@AfroditiMariaZaki here are some initial questions/comments/etc...

I'll give it another read once we deal with this initial batch.

AfroditiMariaZaki committed:
Update tutorials/MD/data/md.mdp

Co-authored-by: Irfan Alibay <IAlibay@users.noreply.github.com>

AfroditiMariaZaki committed:
Update tutorials/MD/data/md.mdp

Co-authored-by: Irfan Alibay <IAlibay@users.noreply.github.com>

AfroditiMariaZaki committed:
Update tutorials/MD/data/md.mdp

Co-authored-by: Irfan Alibay <IAlibay@users.noreply.github.com>

AfroditiMariaZaki committed:
Update tutorials/MD/data/md.mdp

Co-authored-by: Irfan Alibay <IAlibay@users.noreply.github.com>

AfroditiMariaZaki committed:
Update tutorials/MD/prerun/setup/em.mdp

Co-authored-by: Irfan Alibay <IAlibay@users.noreply.github.com>

AfroditiMariaZaki committed:
Update tutorials/MD/prerun/setup/setup/em.mdp

Co-authored-by: Irfan Alibay <IAlibay@users.noreply.github.com>

AfroditiMariaZaki committed:
Update 02_Protein_Visualisation.ipynb

Added instructions on how to observe the protein structure both with VMD and NGLView.

AfroditiMariaZaki committed:
Add files via upload

Updated Notebook 3. Changes suggested by Irfan incorporated.

AfroditiMariaZaki committed:
Delete 03_MD_Setup_Production_Run.ipynb

Replaced with the updated version

AfroditiMariaZaki committed:
Update 03_Running_an_MD_simulation.ipynb

AfroditiMariaZaki committed:
Update 03_Running_an_MD_simulation.ipynb

AfroditiMariaZaki committed:
Delete .md.xtc_offsets.npz

AfroditiMariaZaki committed:
Delete em.log

AfroditiMariaZaki committed:
Delete em.trr

AfroditiMariaZaki committed:
Update 03_Running_an_MD_simulation.ipynb

AfroditiMariaZaki committed:
Update em.mdp

AfroditiMariaZaki committed:
Update md.mdp

_ changed to -

AfroditiMariaZaki committed:
Update 03_Running_an_MD_simulation.ipynb

Removed image of em_potential_energy and instead ask them to observe the decrease in the potential by looking in the .xvg file.

AfroditiMariaZaki committed:
Add instructions on how to observe the decrease in em_potential with vi

AfroditiMariaZaki committed:
Update em.mdp

AfroditiMariaZaki committed:
EM steps corrected

AfroditiMariaZaki committed:
Updated mdp

AfroditiMariaZaki committed:
Updated notebooks and files after review

AfroditiMariaZaki committed:
Updated notebook 03

AfroditiMariaZaki committed:
Updated md submission script

AfroditiMariaZaki committed:
Updated md submission script

AfroditiMariaZaki committed:
Updated em submission script

AfroditiMariaZaki committed:
Added md_fit.xtc that will be used for analysis

AfroditiMariaZaki committed:
Updated Notebook 3 with trj editing commands

AfroditiMariaZaki committed:
Added modified trj

IAlibay commented on tutorials/MD/02_Protein_Visualisation.ipynb:
```diff
@@ -0,0 +1,155 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## <span style='color:darkred'> 2.0 Obtain the protein structure </span>\n",
```
I think we need to make the separation between VMD and NGLview (and also the fact that VMD is optional) more clear here. Otherwise we're going to have a ton of folks that didn't/couldn't download VMD complain .

AfroditiMariaZaki replied:
Have added two sections: VMD (otional) and NGLView

IAlibay commented on tutorials/MD/02_Protein_Visualisation.ipynb:
```diff
@@ -0,0 +1,155 @@
+    "pdbview.gui_style = 'ngl'\n",
+    "\n",
+    "# Define the protein universe\n",
+    "protein=MDAnalysis.Universe(\"./prerun/setup/1hsg.pdb\")\n",
```
If we went through the trouble of downloading a version to the setup directory, why are we pointing this to "prerun"?

Alternatively, we can just use `fetch_mmtf` from MDAnalysis, then you don't even need to have a local PDB.

AfroditiMariaZaki replied:
Fixed

IAlibay commented on tutorials/MD/02_Protein_Visualisation.ipynb:
```diff
@@ -0,0 +1,155 @@
+    "# Import the MDAnalysis and the nglview packages\n",
+    "import MDAnalysis\n",
+    "import nglview\n",
+    "pdbview.gui_style = 'ngl'\n",
```
I'm surprised this isn't throwing a `NameError`. You should be applying this to the view object?

Here's the code that should work here:

```python
pdbview = nv.show_mdanalysis(pdb)
pdbview.gui_style = 'ngl'

# Update the cartoon representation to colour the protein by secondary structure
pdbview.update_cartoon(color="sstruc")

# Change the display a little bit
pdbview.parameters = dict(camera_type='orthographic', clip_dist=0)

# Display the nglview object
pdbview
```

AfroditiMariaZaki replied:
Updated. it works fine now.

IAlibay commented on tutorials/MD/02_Protein_Visualization.ipynb:
```diff
@@ -0,0 +1,158 @@
+{
```
I don't know why there are two notebooks, one named with using american and the other using british english, but probably my comments from the former apply to this one too.

AfroditiMariaZaki replied:
Fixed

IAlibay commented on tutorials/MD/03_Running_an_MD_simulation.ipynb:
```diff
@@ -0,0 +1,315 @@
+    "\n",
+    "## <span style='color:darkred'> 3.1 Connect to the Advanced Research Computing (ARC) </span>\n",
+    "\n",
+    "Open a terminal and remotely connect to ARCUS using ssh (secure shell), using your username:\n",
```
Probably needs something like:

```
Note: if using Windows you will need to connect via MobaXterm as detailed in the setup instructions
```

AfroditiMariaZaki replied:
Added.

IAlibay commented on tutorials/MD/03_Running_an_MD_simulation.ipynb:
```diff
@@ -0,0 +1,315 @@
+    "\n",
+    "`% pwd`\n",
+    "\n",
+    "It should be your home directory (/home/username). This is where you will work from, you will need to create some subdirectories to organise your files. First creatge a directory for your setup and move to it, as it will be your first working directory:\n",
```
creatge -> create

AfroditiMariaZaki replied:
Fixed

IAlibay commented on tutorials/MD/03_Running_an_MD_simulation.ipynb:
```diff
@@ -0,0 +1,315 @@
+    "- The trajectory files have the extension .xtc and .trr, the former does not contain velocity information and coordinates are held at a reduced precision, and so occupies less disk space. However you will need velocities if you want to continue a simulation.\n",
+    "- The .edr file contains the energy information from the trajectory.\n",
+    "- The .mdp file contains the information that was used to setup the actual simulation. Things related to temperature, pressure, how the electrostatics is calculated etc.\n",
+    "- The .tpr file is a binary file that contains all the information needed to perform the actual run (this allows  gromacs to do lots of self-consistency checks to minimize user errors).\n",
```
two empty spaces between "allows" and "gromacs"?

AfroditiMariaZaki replied:
Fixed

IAlibay commented on tutorials/MD/03_Running_an_MD_simulation.ipynb:
```diff
@@ -0,0 +1,315 @@
+    "- The .mdp file contains the information that was used to setup the actual simulation. Things related to temperature, pressure, how the electrostatics is calculated etc.\n",
+    "- The .tpr file is a binary file that contains all the information needed to perform the actual run (this allows  gromacs to do lots of self-consistency checks to minimize user errors).\n",
+    "\n",
+    "### <span style='color:darkred'> 3.2.2 Use GROMACS on ARCUS </span>\n",
```
Use -> Using?

AfroditiMariaZaki replied:
Changed

IAlibay commented on tutorials/MD/03_Running_an_MD_simulation.ipynb:
```diff
@@ -0,0 +1,315 @@
+    "\n",
+    "### <span style='color:darkred'> 3.2.2 Use GROMACS on ARCUS </span>\n",
+    "\n",
+    "In order to use GROMACS on ARCUS, you need to load the already available module:\n",
```
Do we need to make it a little bit clearer as to what a "module" is? Maybe like a line saying that "this makes the software available for you to use, a bit like installing the software on the fly" (which isn't accurate, but at least they get an idea of what it does)?

AfroditiMariaZaki replied:
Added.

IAlibay commented on tutorials/MD/03_Running_an_MD_simulation.ipynb:
```diff
@@ -0,0 +1,315 @@
+    "\n",
+    "As previously discussed in notebook 2, we will use the HIV-1 protease structure (1HSG). It is a homodimer with two chains of 99 residues each. Even though you have already obtained the .pdb file of the protein from the [protein data bank](https://www.rcsb.org/), for this step it is probably easier to just copy it from the shared directory to your working directory. In fact, most of the files ...
+    "\n",
+    "`% cp /home/shared/data/1hsg.pdb .` \n",
```
any luck on a reply from ARCUS about this?

AfroditiMariaZaki replied:
Changed the instruction to guide them to git clone the data repository

IAlibay commented on tutorials/MD/03_Running_an_MD_simulation.ipynb:
```diff
@@ -0,0 +1,315 @@
+    "\n",
+    "If you look at this file (using the graphical editors vi or nedit for example, you should immediately see that it has two chains; A and B:\n",
+    "\n",
+    "`% vi 1hsg.pdb`\\\n",
```
if we make them use `vi` or `nano` we might have to give them an idea about how to exit both (I have a good feeling otherwise we might get 40 emails asking us to tell them how to do `:q`), @RMeli might have opinions here too.

AfroditiMariaZaki replied:
Instructions included on how to exit.

IAlibay commented on tutorials/MD/03_Running_an_MD_simulation.ipynb:
```diff
@@ -0,0 +1,315 @@
+    "\n",
+    "This puts the protein in the centre of the box that is 7 nm x 7 nm x 7 nm and creates the resulting file boxed.gro. As you may remember from the lecture, in this setup we will apply periodic boundary conditions, which means that the box is replicated indefinitely in all three directions. \n",
+    "\n",
+    "Next we need to add water to the system. We can add the water by repeatedly overlaying a small box of water into the system (216 molecules).\n",
```
> by repeatedly overlaying a small box of water into the system (216 molecules).

At least to me this is a little bit confusing, it probably needs to be clarified that this is what `solvate` does?

AfroditiMariaZaki replied:
Rephrased

IAlibay reviewed (commented):
Here are more comments on the changes, it's unclear that I'll have more time for further reviews tomorrow (it depends on how the replacement notebook for the MD Analysis notebook being transplanted here goes...). Possibly @RMeli could review the analysis notebook when it's done?

AfroditiMariaZaki committed:
Unnecessary analysis directory removed

AfroditiMariaZaki committed:
Updated 02_Protein_Visualization.ipynb

AfroditiMariaZaki committed:
Added setup.md

IAlibay commented on tutorials/MD/setup.md:
```diff
@@ -0,0 +1,34 @@
+# Molecular Dynamics (MD) Tutorial
```
Why is this being added? It's going to override the files from #120?

If we want to add this information it should go in the README.md file.

AfroditiMariaZaki replied:
By mistake. Removed now.

AfroditiMariaZaki committed:
setup.md removed

AfroditiMariaZaki committed:
setup.md removed

AfroditiMariaZaki committed:
Updated 01_Introduction.ipynb

AfroditiMariaZaki committed:
Updated 02_Protein_Visualization.ipynb

AfroditiMariaZaki committed:
Updated 03_Running_an_MD_simulation.ipynb

AfroditiMariaZaki committed:
Updated 03_Running_an_MD_simulation.ipynb

AfroditiMariaZaki committed:
Updated 04_Trajectory_Analysis.ipynb

AfroditiMariaZaki committed:
Updated 04_Trajectory_Analysis_Solutions.ipynb

AfroditiMariaZaki committed:
Removed data direcories. They will be in a separate repository

AfroditiMariaZaki committed:
Updated 03_Running_an_MD_simulation.ipynb with some thermostat info

AfroditiMariaZaki commented:
@IAlibay  @philbiggin  Comments addressed and notebooks updated. This is my final version.

Now I am worried that if merged, will the other tutorials that exist in previous versions in my branch mess up everything? Sorry, gihub still confuses me. I am not sure how to handle this.

AfroditiMariaZaki committed:
Updated 03_Running_an_MD_simulation.ipynb with correct selection for gmx energy command

IAlibay committed:
try to fix merge conflict by deleting

IAlibay commented:
> @IAlibay @philbiggin Comments addressed and notebooks updated. This is my final version.
> 
> Now I am worried that if merged, will the other tutorials that exist in previous versions in my branch mess up everything? Sorry, gihub still confuses me. I am not sure how to handle this.

@AfroditiMariaZaki 

The only thing that needed to be handled here was the merge conflict on tutorials/Python/11_AnalysingMD/AnalysingMD_solutions.ipynb

Since the file no longer exists on the `master` branch I just went ahead and deleted it.

There's a lot of files being changed in this PR, so I can't guarantee you might not end up with unecessary files somewhere, but we can deal with this post-merge.

IAlibay reviewed (approved):
In the interest of saving time, I'm going to approve without further review. I can probably spend a bit of time tomorrow morning if you want me to have a read through pre-merge though, otherwise we can go ahead and just merge.

IAlibay commented on tutorials/MD/README.md:
```diff
@@ -1,31 +0,0 @@
-# Oxford Computational Biochemistry Molecular Dynamics Tutorial
```
We're going to need a README.md for this tutorial set (even if it's something basic like what we have in the homology-modelling tutorial). I suggest we just add it post merge of this PR though. As long as the setup.md exists then at least students know what to do.
